omc3/scripts/kmod_lumi_imbalance.py

```diff
@@ -0,0 +1,237 @@
+        opt_cp = copy.copy(opt)
+        for ip in IPS:
+            if not isinstance(opt_cp[ip], (Path, str)):
+                opt_cp[ip] = "(was provided as DataFrame)"
+        save_config(output_path, opt_cp, __file__)
+
+    dfs = _read_and_check_dataframes(opt)
+    df = get_lumi_imbalance_df(**dfs)
+    betastar_x, betastar_y = dfs[list(dfs.keys())[0]].loc[1, [f'{BETASTAR}X{MDL}', f'{BETASTAR}Y{MDL}']].values
+
+    if output_path is not None:
+        tfs.write(
+            output_path / f"{EFFECTIVE_BETAS_FILENAME.format(betastar_x=betastar_x, betastar_y=betastar_y)}{EXT}", 
+            df, 
+            save_index=NAME
+        )
+    
+    return df
+
+
+def _read_and_check_dataframes(opt: DotDict) -> dict[str, tfs.TfsDataFrame]:
+    """
+    Read the given DataFrames if needed, check them for validity and return a dictionary.
+    """
+    dfs = {}
+    for ip in IPS:
+        df = opt.get(ip, None)
+        if df is None:
+            continue
+
+        if isinstance(df, (Path, str)):
+            try:
+                df = tfs.read(df, index=BEAM)
+            except KeyError as e:
+                msg = (
+                    f"Dataframe '{df}' does not contain a '{BEAM}' column."
+                    "You need to run the `kmod_average` script on data for both beams "
+                    "before you can calulate the luminosity imbalance."
+                )
+                raise KeyError(msg) from e
+
+        if BEAM in df.columns:
+            df = df.set_index(BEAM)
+
+        for beam in (1, 2):
+            if beam not in df.index:
+                msg = (
+                    f"Dataframe '{df}' does not seem to contain data per beam."
+                    "You need to run the `kmod_average` script on data for both beams "
+                    "before you can calulate the luminosity imbalance."
+                )
+                raise KeyError(msg)
+
+        dfs[ip] = df
+    
+    if len(dfs) != 2:
+        msg = (
+            "Lumi inbalance can only be calculated for exactly two IPs, "
+            f"but instead {len(dfs)} were given."
+        )
+        raise ValueError(msg)
+
+    return dfs
+
+
+def get_lumi_imbalance_df(**kwargs) -> tfs.TfsDataFrame:
+    """
+    Calculate the effective beta stars and the luminosity imbalance from the input dataframes.
+
+    Args:
+        ipA (tfs.TfsDataFrame): ar`TfsDataFrame` with the averaged results from a kmod analysis, for IP_A.
+        ipB (tfs.TfsDataFrame): a `TfsDataFrame` with the averaged results from a kmod analysis, for IP_B.
+        (Actually, any name that ends with an integer is fine.)
+    
+    Returns:
+        tfs.TfsDataFrame with effective beta stars per IP and the luminosity imbalance added to the header.
+    """
+    df_effective_betas = tfs.TfsDataFrame()
+    for ip, df in kwargs.items():
+        df_effective_betas.loc[ip.upper(), [f'{BETASTAR}', f'{ERR}{BETASTAR}']] = get_effective_beta_star_w_err(df)
```

Comment:
We could have a check that the `ip` var is indeed in the form of `IP[number]` here maybe? Since this is a public function we should make sure we don't do garbage in, garbage out

Reply:
I'm ok with leaving the user the freedom to name it whatever they want. Maybe not all IPs are called "IP\d" in accelerators.